### PR TITLE
[clang/AST] Make it possible to use SwiftAttr in type context

### PR DESCRIPTION
**Pull request by xedin**
Swift ClangImporter now supports concurrency annotations on imported declarations and their parameters/results, to make it possible to use imported APIs in Swift safely there has to be a way to annotate individual parameters and result types with relevant attributes that indicate that e.g. a block is called on a particular actor or it accepts a `Sendable` parameter.

To faciliate that `SwiftAttr` is switched from `InheritableAttr` which is a declaration attribute to `DeclOrTypeAttr`. To support this attribute in type context we need access to its "Attribute" argument which requires `AttributedType` to be extended to include `Attr *` when available instead of just `attr::Kind` otherwise it won't be possible to determine what attribute should be imported.

### CODE REVIEW

File: clang/lib/Sema/SemaType.cpp
Comment (Xazax-hun) [inline clang/lib/Sema/SemaType.cpp:7221]: The formatting looks off here.

File: clang/lib/Sema/SemaType.cpp
Comment (Xazax-hun) [inline clang/lib/Sema/SemaType.cpp:8790]: The formatting might be a bit off here.

File: clang/lib/Sema/SemaType.cpp
Comment (Xazax-hun) [inline clang/lib/Sema/SemaType.cpp:7182]: I don't see a test case covering this scenario, should we add one?

**Review by Xazax-hun — APPROVED** [review]
Overall looks good to me, I have one question and a couple nits inline. I think having access to Attr from AttributedType makes sense, I would not be surprised if this helped some other use cases like Lifetime annotations in Crubit. But please wait for a review from @erichkeane as well.

**Review by erichkeane — COMMENTED** [review]
So I think this patch is doing too much, I'd like to see the changes to `AttributedType` split into a separate patch so we can talk about the `swift_attr` alone.

As far as moving this attribute to be a Type Attribute, that comes with a ton of concerns around templates.  I don't see anything set to instantiate teh arguments, so any value you give to it will not be instantiated (the attribute will as far as I can tell, be instantiated, but with no modifications to its arguments).

I would like to see a some tests (AST and not!) that validate the behavior of this on a template, and as a type being instantiated, to ensure it doesn't get lost in cases where you'd want it to stay.

Also, we need a release note.

File: clang/lib/AST/ASTContext.cpp
Comment (hnrklssn) [inline clang/lib/AST/ASTContext.cpp:5133]: `AttributedType` is uniqued based on `attrKind`, but not `attr`. So calling this function may return a type with a pointer to a different `Attr` (although of the same kind). Is this a problem? Do we want to fully unique `AttributedType` based on it's `Attr*`?

The other option I believe is to fetch the `Attr*` from `AttributedTypeLoc` (which you can get through the `TypeSourceInfo` from `Decl` or `ExplicitCastExpr`) rather than the `AttributedType`, but doing that for generic expressions gets quite hairy.

Comment (xedin) [reply clang/lib/AST/ASTContext.cpp:5133]: I think adding `attr` to profile is a better approach, I'll look into that.

File: clang/include/clang/AST/Type.h
Comment (xedin) [reply clang/include/clang/AST/Type.h:6059]: If I remember correct the `Attr *` is not always there (nullable or something) that's why we need both `Attr *` and `attrKind`

File: clang/lib/Sema/SemaType.cpp
Comment (xedin) [reply clang/lib/Sema/SemaType.cpp:8790]: Oh, it might be tabs vs. spaces :/

File: clang/include/clang/AST/Type.h
Comment (Xazax-hun) [reply clang/include/clang/AST/Type.h:6059]: I see, thanks! In that case it makes sense to me to encode this invariant in the constructor as an assertion (either `attr` is null, or it has the same kind as `attrKind`).

Comment (xedin) [reply clang/include/clang/AST/Type.h:6059]: Sure!

Comment (xedin) [reply clang/include/clang/AST/Type.h:6059]: In fact I added that assert to `getAttribute(...)` method on ASTContext already.

File: clang/lib/Sema/SemaType.cpp
Comment (xedin) [reply clang/lib/Sema/SemaType.cpp:7182]: This actually applies more on the Swift side, I have multiple tests for this in ClangImporter but I'm not sure how to test this here since it doesn't really matter for clang AST where the attribute ends up being located.

File: clang/lib/AST/ASTContext.cpp
Comment (xedin) [reply clang/lib/AST/ASTContext.cpp]: ✅

File: clang/lib/Sema/SemaType.cpp
Comment (xedin) [reply clang/lib/Sema/SemaType.cpp]: ✅

File: clang/lib/Sema/SemaType.cpp
Comment (xedin) [reply clang/lib/Sema/SemaType.cpp]: ✅

File: clang/include/clang/AST/Type.h
Comment (xedin) [reply clang/include/clang/AST/Type.h]: ✅

**Comment by xedin** [comment]
I addressed most of the comments. A few points to add:

- I'm not sure what to make of the comment regarding templates if somebody could point me to the right direction I can make appropriate changes
- I'd need some guidance regarding release note
- I'd prefer to keep AttributedType changes in this PR since they are integral part of the swift_attr change.

**Comment by Xazax-hun** [comment]
I think the interesting template cases are like things like:

```
template<typename T> struct S {
  T foo();
};

S<int> a;
S<SWIFT_SENDABLE int> b;

void f() {
  a.foo();
  b.foo();
}
```

The question is, whether we actually have the type attributes where we expect to have them, or do we lose them?
Hopefully @erichkeane can correct me if I'm wrong. 

That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop. 

For the release notes, you'd want to mention that `swift_attr` can now be applied to types (maybe a small motivation why) in `clang/docs/ReleaseNotes.rst`. It has a section for attribute changes.

**Comment by erichkeane** [comment]
> I think the interesting template cases are like:
> 
> ```
> template<typename T> struct S {
>   T foo();
> };
> 
> S<int> a;
> S<SWIFT_SENDABLE int> b;
> 
> void f() {
>   a.foo();
>   b.foo();
> }
> ```
> 
> The question is, whether we actually have the type attributes where we expect to have them, or do we lose them? Hopefully @erichkeane can correct me if I'm wrong.
> 
> That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop.
> 
> For the release notes, you'd want to mention that `swift_attr` can now be applied to types (maybe a small motivation why) in `clang/docs/ReleaseNotes.rst`. It has a section for attribute changes.

You're correct, the problem with attributed type is that it is going to be lost A LOT every time we canonicalize the type, else they need to be different types, and thus those would be different instantiations.  This is SOMETIMES not problematic (and sometimes even wanted, see vector types), but it IS problematic if you want your attribute to not cause problems being used in containers/etc, since `std::is_same<int, MY_INT_ATTRIBUTED_TYPE>` being 'false' is problematic for many use cases.

**Comment by xedin** [comment]
> That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop.

I don't think we do, we need swift_attr to help us augment Objective-C APIs to make existing frameworks pray well with Swift concurrency. I think C++ Interoperability only uses them in limited set of cases which don't involve templates.

File: clang/lib/Sema/SemaType.cpp
Comment (xedin) [reply clang/lib/Sema/SemaType.cpp:7192]: Done! I remembered why I didn't add that - it required `-fblocks`.

**Comment by xedin** [comment]
Updated release notes and added block tests requested by @Xazax-hun.

**Comment by xedin** [comment]
@erichkeane friendly ping about this, I replied to your last question wondering if it would be okay to land this or should I close since it has been awhile now.

**Comment by erichkeane** [comment]
> > That being said, one question is whether we actually need type attributes to be propagated in this case for Swift's interop.
> 
> I don't think we do, we need swift_attr to help us augment Objective-C APIs to make existing frameworks play well with Swift concurrency. I think C++ Interoperability only uses them in limited set of cases which don't involve templates.

What about ObjC++?  We support that as well, and I'd like for us to not lose that ability too.

File: clang/include/clang/AST/TypeProperties.td
Comment (erichkeane) [inline clang/include/clang/AST/TypeProperties.td:674]: Is there reason to store both `attribute` and `attrKind` here?  Should we just implement `attrKind` as `attribute.getKind` ?

File: clang/lib/AST/Type.cpp
Comment (erichkeane) [inline clang/lib/AST/Type.cpp:4117]: I think it'd be lovely to get these bits back.  Perhaps just remove attrKind from this `ctor` and just implement the above one instead.

**Review by erichkeane — COMMENTED** [review]
I think this would be a great opportunity to save a bit of memory on `AttributedType`, else this is probably ok to me.

File: clang/include/clang/AST/TypeProperties.td
Comment (xedin) [reply clang/include/clang/AST/TypeProperties.td:674]: Yes, Attr is not always available (nullability related attributes only have a kind and don't allocate Attr).

File: clang/lib/AST/Type.cpp
Comment (xedin) [reply clang/lib/AST/Type.cpp:4117]: I don't think we can do that, see https://github.com/llvm/llvm-project/pull/108631/files#r1823004595. @Xazax-hun had a comment to that effect already.

**Review by erichkeane — APPROVED** [review]

**Comment by xedin** [comment]
@erichkeane Do I need to re-run CI? I'd need some help with that since I don't have any permissions :)

**Comment by erichkeane** [comment]
> @erichkeane Do I need to re-run CI? I'd need some help with that since I don't have any permissions :)

IDK how to do that other than push another 'small' patch with a merge commit or something.  You don't HAVE to, but there is of course increased risk of revert if you don't.  I might suggest just committing 1st thing in your morning, then dealing with the build bots during the day as they come up

**Comment by xedin** [comment]
No worries, I can just rebase and push force, that should do it :)

**Comment by xedin** [comment]
@erichkeane CI are green, could you please merge since I don't have permissions?...

**Comment by xedin** [comment]
I think if `noderef` attribute instantiations in your example are considered unique when they appear in different spots then we should have two distinct attributed types but it sounds like both of `noderef` should have one `Attr *` instantiation...

**Comment by bevin-hansson** [comment]
Well, as far as I know Attrs are created on the spot and aren't uniqued. Perhaps that should happen for these kinds of Attrs.

**Pull request merged**
